File: ttt/src/agents/az_nets.py
"""NNs intended for use with alphazero"""

from pathlib import Path

import torch
from torch import nn
import torch.nn.functional as F


# stole this from https://github.com/foersterrobert/AlphaZero
class ResNet(nn.Module):
    def __init__(self, num_res_blocks, num_hidden, device):
        super().__init__()
        self.device = device
        self.num_res_blocks = num_res_blocks
        self.num_hidden = num_hidden

        self.startBlock = nn.Sequential(
            nn.Conv2d(3, num_hidden, kernel_size=3, padding=1),
            nn.BatchNorm2d(num_hidden),
            nn.ReLU(),
        )
        self.backBone = nn.ModuleList(
            [_ResBlock(num_hidden) for _ in range(num_res_blocks)]
        )
        self.policyHead = nn.Sequential(
            nn.Conv2d(num_hidden, 32, kernel_size=3, padding=1),
            nn.BatchNorm2d(32),
            nn.ReLU(),
            nn.Flatten(),
            nn.Linear(32 * 3 * 3, 9),
        )
        self.valueHead = nn.Sequential(
            nn.Conv2d(num_hidden, 3, kernel_size=3, padding=1),
            nn.BatchNorm2d(3),
            nn.ReLU(),
            nn.Flatten(),
            nn.Linear(3 * 3 * 3, 1),
            nn.Tanh(),
        )

        self.to(device)

    def forward(self, x):
        x = self.startBlock(x)
        for resBlock in self.backBone:
            x = resBlock(x)
        policy = self.policyHead(x)
        value = self.valueHead(x)
        return policy, value

    def save(self, path: Path):
        torch.save(
            {
                "num_res_blocks": self.num_res_blocks,
                "num_hidden": self.num_hidden,
                "state_dict": self.state_dict(),
            },
            path,
        )

    @staticmethod
    def load(path: Path, device):
        d = torch.load(path, weights_only=True)
        n = ResNet(d["num_res_blocks"], d["num_hidden"], device)
        n.load_state_dict(d["state_dict"])
        return n


class _ResBlock(nn.Module):
    def __init__(self, num_hidden):
        super().__init__()
        # TODO: are num channels here correct? check ospeil
        self.conv1 = nn.Conv2d(num_hidden, num_hidden, kernel_size=3, padding=1)
        self.bn1 = nn.BatchNorm2d(num_hidden)
        self.conv2 = nn.Conv2d(num_hidden, num_hidden, kernel_size=3, padding=1)
        self.bn2 = nn.BatchNorm2d(num_hidden)

    def forward(self, x):
        residual = x
        x = F.relu(self.bn1(self.conv1(x)))
        x = self.bn2(self.conv2(x))
        x += residual
        x = F.relu(x)
        return x

File: chess/csharp_chess/cschess.agents/AlphaZero/AzNets.cs
using System.Diagnostics;
using cschess.game;
using MoreLinq;
using TorchSharp;
using TorchSharp.Modules;
using static TorchSharp.torch;

namespace cschess.agents.AlphaZero;

/// <summary>
/// An NN that alphazero can use. Given a chess game, outputs move probabilities
/// and value estimate.
/// </summary>
public interface IAzNet
{
    Device Device { get; }
    ICodec Codec { get; }
    IEnumerable<(Dictionary<Move, float>, float)> BatchEval(IEnumerable<IChessGame> games);

    // Raw tensor passthrough for learning
    (Tensor policy, Tensor value) Forward(Tensor states);
    public IEnumerable<Parameter> ModelParams();
}

public class ResNet : IAzNet, IEvaluator
{
    public ICodec Codec { get; } = new Codec4096();
    private readonly ResNetModule _model;
    public Device Device { get; private set; }

    public ResNet(int numResBlocks, int numHidden, Device device)
    {
        _model = new ResNetModule(numResBlocks, numHidden, Codec, device).to(device);
        Device = device;
    }

    public void Train()
    {
        _model.train();
    }

    public void Eval()
    {
        _model.eval();
    }

    public IEnumerable<(Dictionary<Move, float>, float)> BatchEval(IEnumerable<IChessGame> games)
    {
        var gameList = games.ToList();
        var pvs = Pv(gameList).ToList();
        Debug.Assert(pvs.Count == gameList.Count);
        foreach (var pvgs in pvs.Zip(gameList))
        {
            var (pv, game) = pvgs;
            var (p, v) = pv;
            yield return (Codec.Probdist2Dict(p, game), v);
        }
    }

    public (Tensor policy, Tensor value) Forward(Tensor states)
    {
        return _model.forward(states);
    }

    /// <summary>
    /// games to list(probs, value) output
    /// </summary>
    private IEnumerable<(float[], float)> Pv(IEnumerable<IChessGame> games)
    {
        var (logits, values) = Forward(games);
        var parr = logits.softmax(dim: 1).cpu().data<float>().ToArray();
        var varr = values.squeeze().cpu().data<float>().ToArray();
        Debug.Assert(parr.Length == varr.Length * Codec.ActionSize);
        return parr.Batch(Codec.ActionSize).Zip(varr);
    }

    /// <summary>
    /// games to raw model tensor outputs
    /// </summary>
    private (Tensor, Tensor) Forward(IEnumerable<IChessGame> games)
    {
        var arr = Codec.States2Array(games);
        var input = from_array(arr).to(Device);
        return _model.forward(input);
    }

    public IEnumerable<Parameter> ModelParams()
    {
        return _model.parameters();
    }
}

internal sealed class ResNetModule : nn.Module<Tensor, (Tensor, Tensor)>
{
    private readonly Sequential _start;
    private readonly Sequential _backbone;
    private readonly Sequential _policyHead;
    private readonly Sequential _valueHead;

    public ResNetModule(int numResBlocks, int numHidden, ICodec codec, Device device)
        : base("ResNet")
    {
        _start = nn.Sequential(
                ("conv1", nn.Conv2d(8, numHidden, kernel_size: 3, padding: 1)),
                ("bn1", nn.BatchNorm2d(numHidden)),
                ("relu", nn.ReLU())
            )
            .to(device);
        _backbone = nn.Sequential(
                Enumerable.Range(0, numResBlocks).Select(i => new ResBlock(numHidden))
            )
            .to(device);
        _policyHead = nn.Sequential(
                nn.Conv2d(numHidden, 32, kernel_size: 3, padding: 1),
                nn.BatchNorm2d(32),
                nn.ReLU(),
                nn.Flatten(),
                // 8x8 for chess board
                nn.Linear(32 * 8 * 8, codec.ActionSize)
            )
            .to(device);
        _valueHead = nn.Sequential(
                nn.Conv2d(numHidden, 3, kernel_size: 3, padding: 1),
                nn.BatchNorm2d(3),
                nn.ReLU(),
                nn.Flatten(),
                nn.Linear(3 * 8 * 8, 1),
                nn.Tanh()
            )
            .to(device);

        RegisterComponents();
    }

    public override (Tensor, Tensor) forward(Tensor x)
    {
        x = _start.forward(x);
        x = _backbone.forward(x);
        var policy = _policyHead.forward(x);
        var value = _valueHead.forward(x);
        return (policy, value);
    }
}

internal sealed class ResBlock : nn.Module<Tensor, Tensor>
{
    private readonly Conv2d _conv1;
    private readonly Conv2d _conv2;
    private readonly BatchNorm2d _bn1;
    private readonly BatchNorm2d _bn2;

    public ResBlock(int numHidden)
        : base("ResBlock")
    {
        _conv1 = nn.Conv2d(numHidden, numHidden, kernel_size: 3, padding: 1);
        _bn1 = nn.BatchNorm2d(numHidden);
        _conv2 = nn.Conv2d(numHidden, numHidden, kernel_size: 3, padding: 1);
        _bn2 = nn.BatchNorm2d(numHidden);

        RegisterComponents();
    }

    public override Tensor forward(Tensor x)
    {
        var residual = x;
        x = nn.functional.relu(_bn1.forward(_conv1.forward(x)));
        x = _bn2.forward(_conv2.forward(x));
        x += residual;
        x = nn.functional.relu(x);
        return x;
    }
}
